CIP-????/README.md

```diff
@@ -0,0 +1,97 @@
+
+``` Haskell
+mkUniversalScript :: Maybe Datum -> Redeemer -> ScriptContext -> Bool
+mkUniversalScript (Just d) r ctx = ...
+mkUniversalScript Nothing r ctx = ...
+```
+
+### `Data` Specification for `Maybe Datum`
+Before the ledger-script interface passes the datum to the plutus script, the datum will be wrapped in a `Maybe` data type. Seen as type `Data`, this wrapping is:
+
+``` Haskell
+Nothing = Constr 0 []
+Just datum = Constr 1 [datumAsData]
+```
+
+### Using Universal Plutus Scripts
+#### Spending
+If the UTxO being spent is missing a datum, then the `Nothing` constructor will be passed to the plutus script. Otherwise, the datum will be wrapped in the `Just` constructor before being passed to the plutus script.
+
+As an example, to pass the unit datum into the plutus script, the following `Data` would be passed:
+``` Haskell
+Just () = Constr 1 [Constr 0 []]
+```
+
+#### Minting and Staking
+When a universal script is being used for either minting or staking, `Nothing` will automatically be passed into the plutus script.
+
+### Plutus v3
+According to [CIP-0035](https://github.com/michaelpj/CIPs/blob/8e296066c0afc7d2ed46db88eca43f409830e011/CIP-0035/README.md#scripts-in-the-cardano-ledger), what is being suggested here is a change to the ledger-script interface. This means a new Plutus Core ledger language (LL) is required. CIP-0035 states that this requires a hardfork. However, being that the way things are stored on-chain will not change, this should be a smaller hardfork than the past few.
+
+## Rationale: how does this CIP achieve its goals?
+The key idea of this CIP is to address multiple developer concerns as simply as possible. There have been (at least) 5 proposals created so far whose motivations can all be addressed by the `Maybe Datum` ([#364](https://github.com/cardano-foundation/CIPs/pull/364),[#423](https://github.com/cardano-foundation/CIPs/issues/423),[#309](https://github.com/cardano-foundation/CIPs/pull/309),[#310](https://github.com/cardano-foundation/CIPs/pull/310),[#321](https://github.com/cardano-foundation/CIPs/pull/321)). 
+
+With this change, UTxOs would no longer be permanently locked simply for missing a datum. Further, it would now be possible to use the same plutus script for spending, minting, and staking.
+
+### Backwards Compatibility
+Only plutus scripts that use the new plutus ledger language will have their datums handled in this way. All other plutus ledger languages will retain their usage.
+
+### `ScriptArgs` Alternative
+The `ScriptArgs` proposal ([here](https://github.com/cardano-foundation/CIPs/pull/321)) could also address the issues motivating this CIP. The `ScriptArgs` approach could allow for easily implementing future changes such as allowing scripts to take other kinds of arguments. However, all of the details have yet to be ironed out.
```

Comment:
Since this is the major competitor to your proposal, I think you need to say a _lot_ more about why you think yours should be implemented instead!

Reply:
@michaelpj Do you think there is value in implementing the `Maybe Datum` for now while we sort out the `ScriptArgs`? The main difference between the two is that the `ScriptArgs` is easily extendable, but there are still questions around it. The `Maybe Datum` proposal does not take us very far from the existing design. IMO the issues discussed in this CIP are pressing enough to prioritize fixing soon. If we implement the `Maybe Datum` proposal first, we can then take our time sorting out the `ScriptArgs`. If `ScriptArgs` is deemed to be better, how much different is the process of going from our current design to `ScriptArgs` than going from the `Maybe Datum` design to `ScriptArgs`?

Being able to sort out the `ScriptArgs` without the pressure caused by the discussed issues is my main argument for why we should do the `Maybe Datum`. Basically, the idea is to use this CIP as a potential stopgap.

Reply:
No, absolutely not. Changes that require new Plutus ledger languages are very expensive because they require us to support old versions _forever_. I am extremely not keen to add something that we plan to remove later.

Reply:
@michaelpj Understood. Then I don't have an argument for choosing this proposal over the `ScriptArgs` one since I personally believe that, if the `ScriptArgs` can be added safely, it should be used. However, I think there is value in documenting what has been considered at each step of evolution for the design - this has been my main motivation for continuing to steel man this CIP despite thinking that `ScriptArgs` would likely be better. I am totally fine with this proposal getting merged, but then marking its status as inactive once `ScriptArgs` is ready. I'm not sure if the CIP process is meant to be used this way.

Considering that you did something similar with #336, what are your thoughts about this? I'm willing to wait until the `ScriptArgs` proposal is ready and then try arguing against it.